### ISSUE TITLE
[micro-ROS] rclc_executor will fail, if more subscriptions or timers are added.

### ISSUE THREAD

**Issue by JanStaschulat** [gentijo/RosBots.archive#15]
Hi John, thanks for presenting this project in the EWG meeting on 28.02.2023. I made a short review regarding the configuration of the rclc-executor. The following things I'd like to point out: 

- the rclc-executor is configured with two handles (hardcoded) in [`microros_app_base::microros_app_task`](https://github.com/gentijo/RosBots/blob/3df284eb6fce6d93bc55c142e4f5c4415e0b1dca/microros_ws/components/app/ros/microros_app_base.cpp#L138)
- while `microros_app_base::add_ros_subscription` checks for [MAX_SUBSCRIPTIONS](https://github.com/gentijo/RosBots/blob/2f52ed5c48d1f797d448e831a377a323319d3685/microros_ws/components/app/ros/microros_app_base.cpp#L19) =30
- while `microros_timer_mgr::attach()` does [not check the total number of timer handles](https://github.com/gentijo/RosBots/blob/3df284eb6fce6d93bc55c142e4f5c4415e0b1dca/microros_ws/components/ros/microros_timer_mgr.cpp#L5).  
- if you are adding in total more than 2 (subscriptions+timers), this will result in an error: 
  - [`microros_timer_mgr::attach`](https://github.com/gentijo/RosBots/blob/3df284eb6fce6d93bc55c142e4f5c4415e0b1dca/microros_ws/components/ros/microros_timer_mgr.cpp#L5) when calling `rclc_executor_add_timer()` function or
  - [`microros_locomotion_mgr::attach`](https://github.com/gentijo/RosBots/blob/3df284eb6fce6d93bc55c142e4f5c4415e0b1dca/microros_ws/components/ros/microros_locomotion_mgr.cpp#L8) when you are calling the [`rclc_executor_add_subscription`](https://github.com/gentijo/RosBots/blob/3df284eb6fce6d93bc55c142e4f5c4415e0b1dca/microros_ws/components/ros/microros_locomotion_mgr.cpp#L37) function.

You need to align this, if you'll add more subscriptions or timers in the future. You have to configure the rclc-executor with the maximum number of handles that the application is going to use. So as a quick fix, 
- when adding a new timer, check for maximum number of timers, e.g. MAX_TIMERS
- in the function [`rclc_executor_init`](https://github.com/gentijo/RosBots/blob/3df284eb6fce6d93bc55c142e4f5c4415e0b1dca/microros_ws/components/app/ros/microros_app_base.cpp#L138) change the [`number_of_handles`](https://github.com/ros2/rclc/blob/3b4ff9e46ad63ca08a1131760a86b0f42c2bba63/rclc/include/rclc/executor.h#L147) from 2 to (MAX_SUBSCRIPTIONS+MAX_TIMERS)
- be aware, that if you are using `max_handles=30`, this is much more than the default buffers in XRCE-DDS middleware. See the [note](https://github.com/ros2/rclc/blob/3b4ff9e46ad63ca08a1131760a86b0f42c2bba63/rclc/include/rclc/executor.h#L106) here as well.

**Comment by gentijo** [comment]
Hi @JanStaschulat thank you so much for mentioning this bug, totally align with you and will put this on the list
My approach to this project has had a lot of compartmentalized discovery where I hack/slash on a new problem, then every once in a while I used a design tool to reverse engineer my code, stand back and look at the bigger picture then re-moodel to make the code more structured.  The hack/slash approach leads to this kind of bug. 

If interested here is the latest version of the picture, https://gentijo.github.io/rosbots_design/. 
I plan to bring this design docs into the repo once it stabilizes a bit more 
I used this tool, https://sparxsystems.com/ and the model is in the design folder of the project root. 
EA runs great on a Mac using crossover from codeweavers (managed Wine) or Linux using Wine.

**Comment by JanStaschulat** [comment]
@gentijo It's my pleasure to support you in your project using micro-ROS. Thanks for sharing the design document. It really helps to understand the class structure of the project. 

I just have another question regarding your[ sleep](https://github.com/gentijo/RosBots/blob/2f52ed5c48d1f797d448e831a377a323319d3685/microros_ws/components/app/ros/microros_app_base.cpp#L193). What is your intention? 

I did not find the definition of `portTICK_PERIOD_MS`, but I guess you want to wait for some constant time before calling `rclc_executor_spin_some` again with the intension of a periodic activation. Why is this necessary? This wait statement is also misleading, `spin_some` is NOT called periodically with the same time period: because the execution time of `spin_some` depends on the sum of execution times of all processed callbacks, which may vary in each iteration (sometimes a message has arrived - sometimes not ....).

I just wanted to mention, that there is a periodic spin function as well in micro-ROS: [`rclc_spin_period()`](https://github.com/ros2/rclc/blob/b46dba3daab335967a62b338dd30c51f0a5d8261/rclc/include/rclc/executor.h#L825). However, I don't think, it should be used in this application. I'd rather propose just to use the `rclc_executor_spin` function, like in [this example](https://github.com/ros2/rclc/blob/b46dba3daab335967a62b338dd30c51f0a5d8261/rclc_examples/src/example_executor.c#L163). The default timeout is one second. If necessary, you could configure the wait_timeout with [rclc_executor_set_timeout](https://github.com/ros2/rclc/blob/b46dba3daab335967a62b338dd30c51f0a5d8261/rclc/include/rclc/executor.h#L169).

In ROS 2, you typically specify a really long timeout, typically one second, so the spin_function waits on the DDS layer for new events (new message, timer ready). It is reactive - so to say. There is nothing else to do - so this blocking wait is perfectly fine. If you want periodic events, use a timer. 

Be aware, the higher frequency you call `spin_some` function (e.g. shorter timeout in `spin_some`), the greater overhead the ROS middleware (ROS 2 as well as micro-ROS, as RCL is the common layer) will have. Internally structures are re-build every time spin_some is called. See also this [documentation about Executors in ROS 2](https://docs.ros.org/en/humble/Concepts/About-Executors.html).

**Comment by gentijo** [comment]
This code was based on example code, either the ESP32 example or TurtleSim I used both of those project to build my first TurtleBot. Initially it would spin_some, then usleep(n)  which ran into an issue where MPython would not run, thinking about it now I think it is because the MRos task is running at a higher priority so spin_some + usleep never gave up control to lower priority tasks. When I change it to vtask_delay, then it was able to give control to the other tasks. I am not partial to spin_some + delay, but it needs to multitask well with all tasks.
